hal/src/nRF52840/ble_hal.cpp

```diff
@@ -1321,208 +1280,188 @@ ble_gap_scan_params_t BleObject::Observer::toPlatformScanParams() const {
+            if (observer->scanParams_.active && report.type.scannable && !report.type.scan_response) {
+                // Advertising data packet, scan response data is expected.
+                if (observer->getPendingResult(newAddr) != nullptr) {
                     observer->continueScanning();
-                    return;
-                }
-                uint8_t* advData = (uint8_t*)observer->observerScannedDataPool_.alloc(advReport.data.len);
-                if (!advData) {
-                    LOG(ERROR, "Allocate memory for scanned advertising data failed.");
-                    observer->stopScanning();
-                    return;
+                    break;
                 }
-                //LOG_DEBUG(TRACE, "Cached adv_data address: %u, len: %u", (unsigned)advData, advReport.data.len);
-                hal_ble_gap_on_scan_result_evt_t result = {};
-                result.type.connectable = advReport.type.connectable;
-                result.type.scannable = advReport.type.scannable;
-                result.type.directed = advReport.type.directed;
-                result.type.extended_pdu = advReport.type.extended_pdu;
-                result.rssi = advReport.rssi;
-                result.peer_addr.addr_type = newAddr.addr_type;
-                memcpy(result.peer_addr.addr, newAddr.addr, BLE_SIG_ADDR_LEN);
-                result.adv_data_len = advReport.data.len;
-                memcpy(advData, advReport.data.p_data, advReport.data.len);
-                result.adv_data = advData;
-                if (observer->addPendingResult(result) != SYSTEM_ERROR_NONE) {
-                    observer->stopScanning();
-                    return;
-                }
-                observer->continueScanning();
-                return;
             }
-            // It is a scan response.
-            hal_ble_gap_on_scan_result_evt_t* result = observer->getPendingResult(newAddr);
-            if (!result) {
-                observer->continueScanning();
-                return;
+            ble_evt_t* observerEvent = (ble_evt_t*)BleObject::getInstance().dispatcher()->allocEventData(sizeof(ble_evt_t));
+            if (!observerEvent) {
+                LOG(ERROR, "Allocate memory for BLE event failed.");
+                observer->stopScanning();
+                break;
             }
-            if (advReport.data.len > 0) {
-                uint8_t* srData = (uint8_t*)observer->observerScannedDataPool_.alloc(advReport.data.len);
-                if (!srData) {
-                    LOG(ERROR, "Allocate memory for scan response data failed.");
+            // Copy the SoftDevice event.
+            memcpy(observerEvent, event, sizeof(ble_evt_t));
+            ble_gap_evt_adv_report_t& advReport = observerEvent->evt.gap_evt.params.adv_report;
+            if (event->evt.gap_evt.params.adv_report.data.len > 0) {
+                advReport.data.p_data = (uint8_t*)BleObject::getInstance().dispatcher()->allocEventData(advReport.data.len);
```

Comment:
No, it can not. The memory allocated for `ble_evt_t` will be freed once the event is processed from thread. But the memory allocated for the advertising data will retained until this piece of data is dispatched to user application.